doc/designs/quic-design/quic-ackm.md

```diff
@@ -0,0 +1,412 @@
+} QUIC_ACKM_TX_PKT;
+
+int QUIC_ACKM_on_tx_ack_packet(QUIC_ACKM *ackm, const QUIC_ACKM_TX_PKT *pkt);
+```
+
+### On RX Datagram
+
+This must be called whenever a datagram is received. A datagram may contain
+multiple packets, and this function should be called before the calls to
+`QUIC_ACKM_on_rx_packet`.
+
+The primary use of this function is to inform the ACK manager of new credit to
+the anti-amplification budget. Packet and ACK-frame related logic are handled
+separately in the subsequent calls to `QUIC_ACKM_on_rx_packet` and
+`QUIC_ACKM_on_rx_ack_frame`, respectively.
+
+Returns 1 on success.
+
+```c
+int QUIC_ACKM_on_rx_datagram(QUIC_ACKM *ackm, size_t num_bytes);
+```
+
+### On RX Packet
+
+This must be called whenever a packet is received. It should be called after
+`QUIC_ACKM_on_rx_datagram` was called for the datagram containing the packet.
+
+Returns 1 on success.
+
+```c
+typedef struct quic_ackm_rx_pkt_st {
+    /* Thje packet number of the received packet. */
+    uint64_t pkt_num;
+
+    /* The time at which the packet was received. */
+    OSSL_TIME time;
+
+    /*
+     * One of the QUIC_PN_SPACE_* values. This qualifies the pkt_num field
+     * into a packet number space.
+     */
+    unsigned int pkt_space :2;
+
+    /* 1 if the packet has one or more ACK-eliciting frames. */
+   unsigned int is_ack_eliciting :1;
+} QUIC_ACKM_RX_PKT;
+
+int QUIC_ACKM_on_rx_packet(QUIC_ACKM *ackm, const QUIC_ACKM_RX_PKT *pkt);
+```
+
+### On RX ACK Frame
+
+This must be called whenever an ACK frame is received. It should be called
+after any call to `QUIC_ACKM_on_rx_packet`.
+
+The ranges of packet numbers being acknowledged are passed as an argument.
+`pkt_space` is one of the `QUIC_PN_SPACE_*` values, specifying the packet number
+space of the containing packet. `rx_time` is the time the frame was
+received.
+
+This function causes `on_acked` callbacks to be invoked on applicable packets.
+
+Returns 1 on success.
+
+```c
+typedef struct quic_ackm_ack_range_st {
+    /*
+     * Represents an inclusive range of packet numbers [start, end].
+     * start must be <= end.
+     */
+    QUIC_PN start, end;
+} QUIC_ACKM_ACK_RANGE;
+
+typedef struct quic_ackm_ack {
+    /*
+     * A sequence of packet number ranges [[start, end]...].
+     *
+     * The ranges must be sorted in descending order, for example:
+     *      [ 95, 100]
+     *      [ 90,  92]
+     *      etc.
+     *
+     * As such, ack_ranges[0].end is always the highest packet number
+     * being acknowledged and ack_ranges[num_ack_ranges-1].start is
+     * aalways the lowest packet number being acknowledged.
+     *
+     * num_ack_ranges must be greater than zero, as an ACK frame must
+     * acknowledge at least one packet number.
+     */
+    const QUIC_ACKM_ACK_RANGE  *ack_ranges;
+    size_t                      num_ack_ranges;
+
+    OSSL_TIME                   delay_time;
+    uint64_t                    ect0, ect1, ecnce;
```

Comment:
What is this?

Reply:
These are the ECN Count fields in the ACK frame. See https://www.rfc-editor.org/rfc/rfc9000.html#name-ack-frames